src/BuiltInTools/HotRestart/Program.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.Loader;
+
+namespace Microsoft.Extensions.DotNetDeltaApplier
+{
+    internal static class Program
+    {
+        public static void Main(string[] args)
+        {
+            const string envName = "_DOTNET_WATCH_APP_ASSEMBLY_PATH";
+            var appAssemblyPath = Environment.GetEnvironmentVariable(envName) ??
+                throw new InvalidOperationException($"No value found for environment variable '{envName}'.");
+
+            var observer = new Observer();
+            using var _ = DiagnosticListener.AllListeners.Subscribe(observer);
+
+            // Microsoft.Extensions.DotNetApplier's startup hook emulates Control-C to terminate the running app when
+            // it sees changes to Startup.cs / Program.cs during a hot reload session.
+            // We need to differentiate between the emulated Control-C and a Control-C performed by the user on the terminal.
+            // We use a DiagnosticListener to signal any time the StartupHook emulates Control-C, so that we know
+            // when it is not-signaled, it must be a real Control-C.
+            while (observer.EmulatedControlC)
+            {
+                observer.EmulatedControlC = false;
+
+                var currentLoadContext = new HotRestartLoadContext();
+                var mainAssembly = currentLoadContext.LoadFromAssemblyPath(appAssemblyPath);
+                // See https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/AssemblyLoadContext.ContextualReflection.md for details.
+                using var _ = AssemblyLoadContext.EnterContextualReflection(mainAssembly);
+
+                mainAssembly.EntryPoint!.Invoke(null, new object[] { args });
+                currentLoadContext.Unload();
+            }
+        }
+
+        internal sealed class HotRestartLoadContext : AssemblyLoadContext
+        {
+            public HotRestartLoadContext() :
+                base(isCollectible: true)
+            {
+            }
+        }
+
+        private class Observer : IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object?>>
```

Comment:
Small comment describing this as well and what it's for

Reply:
There's a bit of comment describing what this does before the while loop. I think that does a decent job?